Alert Component

rmkraus opened the pull request:
**Description:**
I wanted to get some feedback on this idea before I got too far down the rabbit hole. Tests and documentation would follow. The alert component is designed to check the state of a specified entity. If the entity value becomes equal to the specified value, a repeating alert will be thrown using a notifier. The alert will repeat at a regular interval specified in the config in minutes. The alert can be acknowledged to be silenced. The alerts stop when the entity value is no longer in the specified alert state. 

I use this to tell me if my garage door has been left open for 30 minutes. I get an alert every 30 minutes that the garage is open. I can go onto the frontend and acknowledge the alert so that I don't receive any more alerts until the next time the garage is left open. I use the skip_first option so that I am not notified immediately after the door opens, but only after 30 minutes.

This could also be used to monitor the battery level of devices. A template_sensor could be used to evaluate an expression with greater than, less than, etc. Any condition that you may want to absolutely be aware of, this will do it. It will bug you until you acknowledge the event.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1787

**Example entry for `configuration.yaml` (if applicable):**
```yaml
alert:                                                                         
    garage_door:                                                                  
          name: Garage is open                                                 
          entity_id: input_boolean.garage_door                                 
          state: 'on'   # Optional, 'on' is the default value                                                        
          repeat: 
              - 15
              - 30
              - 60                                                            
          can_acknowledge: True  # Optional, default is True
          skip_first: true  # Optional, false is the default                                                     
          notifiers:                                                           
              - ryans_phone
              - kristens_phone
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

rmkraus committed:
Initial addition of the alert component.

The alert component is designed to check the state of a specified
entity. If the entity value becomes equal to the specified value, a
repeating alert will be thrown using a notifier. The alert can be
acknowledged to be silenced. The alerts stop when the entity value is
no longer in the specified alert state.

rmkraus committed:
Added async support to alert component.

Added async callbacks for services and changed the component setup to
be asynchronous.

rmkraus commented:
I added some more async support. I think everything should be asynchronous now, but I'm not super confident with asyncio. Let me know if I missed anything.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+            alert.get(CONF_STATE, STATE_ON), alert[CONF_REPEAT],
+            alert.get(CONF_SKIP_FIRST, False),
+            alert[CONF_NOTIFIERS])
+        ALL_ALERTS[entity.entity_id] = entity
```
Don't use globals. You can just have this live inside this method. If you really need a global use `hass.data`, which is a dict to store global data.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+            alert[CONF_NOTIFIERS])
+        ALL_ALERTS[entity.entity_id] = entity
+
+    descriptions = {}
```
Please add service descriptions

rmkraus replied:
Roger. I just wanted to get the general idea out there before descriptions, documentation, and tests. These will come soon.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+                yield from alert.async_turn_off()
+
+    for alert in alerts.get(CONF_ENTITIES):
+        entity = yield from hass.loop.run_in_executor(
```
You should be able to add these without the executor. Just remove the update_ha_state call from the constructor and you can do it async.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        event.async_track_state_change(hass, entity,
+                                       self.watched_entity_change)
+
+        self.update_ha_state()
```
remove this.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        self._firing = False
+        self._ack = False
+        self._cancel = None
+        self.entity_id = DOMAIN + '.%s' % slugify(name)
```
please define an ENTITY_ID_FORMAT at the top of the file

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+    @property
+    def hidden(self):
+        """Hide the alert when it is not firing."""
+        return not self._firing
```
I'm surprised by this. How can one turn the alert off when it is not firing?

rmkraus replied:
You can't turn off the alert when it is not firing. This hides the entity when it is not firing.

rmkraus replied:
Wait... I think I just figured out your question. My bad.

This is an alarm, it cannot be silenced if it is not firing. Because there is nothing to silence. It is already silent.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        """Hide the alert when it is not firing."""
+        return not self._firing
+
+    def watched_entity_change(self, entity, from_state, to_state):
```
Should by either async or callback.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        if to_state.state != self._alert_state and self._firing:
+            return self.end_alerting()
+
+    def begin_alerting(self):
```
Should be either coroutine or callback

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+
+        self.update_ha_state()
+
+    def end_alerting(self):
```
Idem

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        self._firing = False
+        self.update_ha_state()
+
+    def _schedule_notify(self):
```
Idem

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+
+    def turn_on(self):
+        """Unacknowledge alert."""
+        _LOGGER.info('Reset Alert: %s', self._name)
```
that's not an info worthy event. Maybe, debug.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+
+    def turn_off(self):
+        """Acknowledge alert."""
+        _LOGGER.info('Acknowledged Alert: %s', self._name)
```
that's not an info worthy event. Maybe, debug.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+                    'notify', target, {'message': self._name})
+        self._schedule_notify()
+
+    def turn_on(self):
```
Please implement `async_turn_on`

rmkraus replied:
Done.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+        self._ack = False
+        self.update_ha_state()
+
+    def turn_off(self):
```
Please implement `async_turn_off`

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,233 @@
+    def watched_entity_change(self, entity, from_state, to_state):
+        """Determine if the alert should start or stop."""
+        _LOGGER.debug('Watched entity (%s) has changed.', entity)
+        if to_state.state == self._alert_state and not self._firing:
```
This is quite limited. For example it won't work with numeric values ?

rmkraus replied:
Oh but it does! If this component is paired with a template sensor, it can be used to trigger based off almost anything!

I figured it was best not to reinvent the wheel and, rather, just rely on the component that already exists.

balloob commented:
This will require tests to be written. Also since it is an automation component will have to be fully async.

rmkraus committed:
Improved async support in alert component.

rmkraus commented:
Thanks for the feedback @balloob. I hadn't kept up with all of the asyncio changes, so I knew I would be missing a thousand things. I appreciate you pointing them out.

I will write service descriptions, docs, and tests now. I wanted to wait to see if I would get another Dexter gif before I went all out.

rmkraus commented:
@kellerza mentioned an interesting idea on Gitter. He suggested having this call a script instead of calling the notifier directly. Does anyone else have any thoughts on this? It would be a bit more flexible, but I fear it may make configuration more involved. I'm torn.

kellerza commented:
@rmkraus the idea was actually a bit different

In my view `alert` should *only* be a set of services that you can call from a script, similar to `notify`. For example `alert.alert` would start the message & `alert.cancel` would stop the message (different from acknowledge)

So that mean `alert` don't monitor entities, that is done via automations & alerts triggered via scripts. This opens up possibilities of all the script's matching and *conditions* etc, which you currently bypass by doing the entity monitoring yourself. A little bit more config, but more in line with containing functionality in components.

```yaml
alert:
  garage_alert:
    message: You left the garage door open for {}
    repeat: 30
    skip_first: true  # Optional, false is the default
    notifiers:
       - ryans_phone
       - kristens_phone

automation:
  - alias: "Action - Garage Door Open"
    trigger:
      platform: state
      entity_id: binary_sensor.garage_door
      from: 'off'
      to: 'on'
    action:
      - condition: state
        entity_id: group.all_devices
        state: 'not_home'
      - service: alert.alert
        entity_id:
          - alert.garage_alert

  - alias: "Action - Garage Door Close"
    trigger:
      platform: state
      entity_id: binary_sensor.garage_door
      from: 'on'
      to: 'off'
    action:
      - service: alert.cancel
        entity_id:
          - alert.garage_alert
```

Otherwise we'll end up duplicating `automation` triggers & conditions in `alert`...

rmkraus commented:
Thank you for clarifying your suggestion. My apologies for mangling it. I'm inclined to keep the alert component the way it is. I am also not planning on duplicating all of the platforms from automations in the alert component. The way I would envision this, more in-depth conditions would be implemented using template binary sensors. 

```yaml
binary_sensor:
    - platform: template
      sensors:
        motion_battery_low:
          value_template: '{{ states.sensor.motion.attributes.battery < 15 }}'
          friendly_name: 'Motion battery is low'

alert:                                                                         
    entities:                                                                  
        - name: Motion Battery is Low                                                 
          entity_id: binary_sensor.motion_battery_low                                 
          state: 'on'   # Optional, 'on' is the default value                                                        
          repeat: 30                                                                                                 
          notifiers:                                                           
              - ryans_phone
              - kristens_phone
```

This allows for all the complication that could possibly be desired in the alerts. Similarly, you could have an automation toggle an input boolean if that were preferable. Your example would then be:

```yaml
alert:
  garage_alert:
    entity_id: input_boolean.garage_door     
    state: 'on'
    message: You left the garage door open
    repeat: 30
    skip_first: true  # Optional, false is the default
    notifiers:
       - ryans_phone
       - kristens_phone

input_boolean:
    garage_open:
        name: Garage is Open
        initial: off

automation:
  - alias: "Action - Garage Door Open"
    trigger:
      platform: state
      entity_id: binary_sensor.garage_door
      from: 'off'
      to: 'on'
    action:
      - condition: state
        entity_id: group.all_devices
        state: 'not_home'
      - condition: state
        entity_id: input_boolean.garage_open
        state: 'on'

  - alias: "Action - Garage Door Close"
    trigger:
      platform: state
      entity_id: binary_sensor.garage_door
      from: 'on'
      to: 'off'
    action:
      - condition: state
        entity_id: input_boolean.garage_open
        state: 'off'
```

I wouldn't recommend this over the template binary sensor, but it would certainly work just fine for anyone that may prefer it. With the current implementation, you can still get all of the functionality you want, but there is also an added benefit that the alert component is functional enough to stand alone by itself. For simple checks, which I presume would envelope 80% of the use cases, the alert component is functional enough by itself. For the last 20% of use cases, the template binary sensor helps fill in the gaps.

That's my thinking.

pvizeli commented:
I don't like it. I think the best solution for handle this is to modify `automation/state.py` to allow repeate state time listener from `for` option and all this can done with a simple automation without new component.

rmkraus commented:
Allowing an automation to repeat would, unfortunately, not offer the same functionality. 

Fundamentally, an alert is something that can be triggered and then fires until it is acknowledged or silenced. It will then not fire again until it is resolved and fires again. Think of a fire alarm or other software like Zabbix or Nagios.

If automations were altered to allow them to repeat, you could disable the automation to simulate acknowledgement. However, once the trigger goes false, it will not enable itself again and you will miss future alerts. Repeating could be useful on automations, but that is not an alert.

In my mind, because the functionality of alerts is fundamentally different from automations, it is best to add a light weight component to handle the new functionality rather than augmenting an existing and stable component to handle something it wasn't made to do.

rmkraus committed:
Added service descriptions and toggle to alert.

The alert service was registered for the alert component, but the
entities did not support toggling. Toggle and Asynchronous Toggling was
added to the entities.

Service descriptions for the alert component were added to the
service.yaml file. These are now loaded asynchronously by the alert
component during setup.

rmkraus committed:
Asyncio related fixes to alert component.

1) Ensured that all comments correctly identified method action.
2) Added synchronous methods to compliment async methods.
3) Ensured that all alert entity values were updated once setup was
complete. Alert entities were not appearing in HASS until the had been
triggered the first time. This fixes that.
4) Converted _notify and _schedule_notify to coroutines.

rmkraus committed:
First set of tests for the alert component.

pvizeli commented:
What is worng about that?
```
automation:
  - alias: "Action - Garage Door Open"
    trigger:
      platform: state
      entity_id: binary_sensor.garage_door
      from: 'off'
      to: 'on'
      for: 30
      repeate: true

    action:
      - condition: state
        entity_id: group.all_devices
        state: 'not_home'
     - service: notify.group
       data:
       message: 'Garage is open!'
```

You can safe a lot of good and complexity...

kellerza commented:
@rmkraus you could probably include the `condition` into the template sensor (say don't alert on Saturdays when its open for the whole day etc)

The only other two features that would really make it useful: 
- Backoff, probably implemented as a list: [15,30,60] would alert at 15, 30, 1hr, 2hr, 3hr, 4hr .... The last one in the list is the final interval.
- Option to disable acknowledgement. If the garage door is open, you should really close it to cancel the alert, not be able to acknowledge it. With backoff, you can minimize the amount of messages.

rmkraus commented:
@pvizeli That is possible, but I think it starts getting out of hand pretty quickly. Once all the potential tweaks to automations are complete to really handle this behavior, I think it would provide a less manageable code base.

@kellerza I agree, I think this + a template sensor gives you nearly all the functionality one could want. I have an idea of how backoff could be implemented. I was toying around with the idea of making the alert interval increasingly smaller as the alert went on. Basically the same as backoff, just the opposite. I'll also play with your second suggestion. Should have an update tonight.

rmkraus committed:
Added back off and ack disabling to alerts.

1) A back off factor was added. The delay between alerts will be
multiplied by this factor after every alert. There are max and min
bounds set of 1 minute and 24 hours.

2) Added a can_acknowledge option to the config. This allows the user
to disable alert acknowledgement by never unhiding the entity card.

rmkraus commented:
@kellerza I added the backoff feature and disabling of acknowledgments that you mentioned.

The backoff feature works a bit differently. It is a factor. Every time the notifier is called, the current delay value will be multiplied by the backoff to calculate the next delay time. The first pause is the specified time. This way, the delay can either be set to increase or decrease frequency. It is dead simple and pretty flexible. There are max and min delays of 1 day to 1 minute. The delays will not be allowed to escape those bounds.

Acknowledgments are disable by never un-hiding the entity. They can still be acknowledged with the service call.

rmkraus committed:
Ensure the first delay is predictable.

Forced the first message delay in the alert component to be the user
specified value. Before, if the first message was sent, the first delay
would be the specified delay multiplied by the back off factor.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+    hass.services.call(DOMAIN, SERVICE_TOGGLE, data)
+
+
+def setup(hass, config):
```
This is not needed.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+
+def turn_off(hass, entity_id):
+    """Acknowledge alert."""
+    data = {ATTR_ENTITY_ID: entity_id}
```
Instead of implementing each method twice, use `fire_callback_threadsafe`

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+    for alert in alerts.get(CONF_ENTITIES):
+        entity = Alert(hass, alert[CONF_NAME],
+                       alert[CONF_ENTITY_ID],
+                       alert.get(CONF_STATE, DEFAULT_STATE),
```
move the defaults into the `voluptuous` schema

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+                       alert[CONF_NOTIFIERS],
+                       alert.get(CONF_CAN_ACK, DEFAULT_CAN_ACK),
+                       alert.get(CONF_BACKOFF, DEFAULT_BACKOFF))
+        all_alerts[entity.entity_id] = entity
```
How are you going to avoid that you are not overwriting an entity here?

To avoid these conflicts, for example the group component has the user specify the object id as a key in a dict:

```
some_alert:
  can_ack: false
```

rmkraus replied:
This has been fixed. The configuration is now more like groups. Good catch.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+        """Determine if the alert should start or stop."""
+        _LOGGER.debug('Watched entity (%s) has changed.', entity)
+        if to_state.state == self._alert_state and not self._firing:
+            return self.begin_alerting()
```
Don't return a coroutine but instead use `yield from`

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+    @asyncio.coroutine
+    def begin_alerting(self):
+        """Begin the alert procedures."""
+        _LOGGER.info('Begining Alert: %s', self._name)
```
"Beginning"

balloob replied:
Also, this can be `debug` level because the state change will already be logged as 'info' to indicate something happened.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+    @asyncio.coroutine
+    def end_alerting(self):
+        """End the alert procedures."""
+        _LOGGER.info('Ending Alert: %s', self._name)
```
Same.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+        self._cancel()
+        self._ack = False
+        self._firing = False
+        yield from self.async_update_ha_state()
```
Since no one is wiating for the result of this, you can call `hass.async_add_job(self.async_update_ha_state())`. Then this method and `watched_entity_change ` can become a `@callback`

rmkraus replied:
I tried changing them to callbacks, but it doesn't let me call self._cancel().

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+                    'notify', target, {'message': self._name})
+        yield from self._schedule_notify()
+
+    def turn_on(self):
```
This will not be called and can be removed.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+        self._ack = False
+        yield from self.async_update_ha_state()
+
+    def turn_off(self):
```
Same

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,294 @@
+        self._ack = True
+        yield from self.async_update_ha_state()
+
+    def toggle(self):
```
Same.

kellerza commented:
@rmkraus does that mean that the delay will keep on increasing?

So delay 15 and multiplier 2 will give: 15, 45, 1:30, 3:00, 6:00

There should be an upper limit - a day is too long, then you'll forget the garage door open for the night. That is why I prefer the delay as a list, nice defined intervals, no math, and a specific (but still useful end result)

rmkraus commented:
Yeah, 2 would be a crazy high factor. A much more appropriate factor for a fast decay would be 1.5:

A 15 minute delay would be: 15, 22.5, 33.75, 50.625, 75.9375, 113.90625

This would have your first alert come in only 15 minutes after the event (if you tell it to skip the first), you'll get two alerts in the first hour of the garage being open, one alert in the next hour, one alert in the hour after that, then the alerts become more than an hour apart. That is pretty close to the behavior you indicated and it is much simpler programatically. That is why I like it.

rmkraus committed:
Cleanup to alert component.

1) Removed unused synchronous methods from the alert component.
2) Changed the alert config format to avoid duplicate entity ids.
3) Ensured all sync functions call their async counterpart instead of
reimplementing the functionality.

rmkraus committed:
Added tests for the alert component.

1) Updated the alert config format in the tests.
2) Added a test for the alert back off feature.
3) Added a test for disabling alert acknowledgments.

rmkraus committed:
Removed import in alert component.

Removed the import of CONF_ENTITIES in alert.py because it is no longer
used.

kellerza commented:
Programatically maybe, but you realise you just used "it's simple" and a series of intervals containing up to 6 decimal places in the same paragraph?

And 2 hours borders on being able to miss it, the next one even worse!

If you dont like a list of `intervals` (no additional keys in config) can you please add a `maximum interval` to your idea of `factor`?

With `ensure_list` a list or single value can be configured, so very simple and predictable from a user perspective

rmkraus commented:
Why, yes I do. 😄 The exact times between intervals need not be totally predictable to the person receiving the alert. Because no one in the history of ever has ever sat around waiting for an alert. If they were doing that, there would be no need for the alert in the first place. 

I can add specifying the min and max interval. After that, I'm hesitant to add too much more. This component is already satisfying too many edge cases.

rmkraus committed:
Added custom min/max delays in alerts.

Added the ability to specify the min and max delay limits for an alert
in the configuration. Also added tests for this feature.

pvizeli commented:
Now it have good feature :+1:

kellerza commented:
@rmkraus to be honest the messages will drive me nuts. I can guarantee you that 80% of HASS users have enough OCD that this will bother them (might even be higher)

Taking your example of 1.5x, you will probably get the following alerts:
- "Garage door open for 15 min"
- "Garage door open for 37.5min" 
- "Garage door open for 71.25 min"
- "Garage door open for 121.875 min"
- "Garage door open for 197.8125 min"

That 4th & 5th message will drive people to learn Python to PR a change.

Luckily you can now get sane messages if `backoff` is 2, and `max` of one or two hours with a `repeat` interval of 15. But that is one of the few use cases I could find with your implementation of 4 configuration keys. (btw, what is the use case of `min`?). 

Its still not as nice as my current imp implementation that I've used for the last couple of years [here](https://github.com/kellerza/electric-imp/blob/master/device/PulseRepeat.class.nut#L38) with `[15,15,30,60]` you get very clean alerts:
- "Garage door open for 0:15"
- "Garage door open for 0:30"
- "Garage door open for 1:00"
- "Garage door open for 2:00"
- "Garage door open for 3:00"
....etc

pvizeli commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+
+def turn_on(hass, entity_id):
+    """Reset the alert."""
+    run_callback_threadsafe(hass.loop, async_turn_on, hass, entity_id)
```
Call only the service with `hass.services.call`. It will be adapt in core.

balloob replied:
Ha, I literally requested the opposite. We want to implement each logic only once and thus call the async version of the same method.

pvizeli replied:
ok

pvizeli commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+
+def turn_off(hass, entity_id):
+    """Acknowledge alert."""
+    run_callback_threadsafe(hass.loop, async_turn_off, hass, entity_id)
```
same too

pvizeli commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+
+def toggle(hass, entity_id):
+    """Toggle acknowledgement of alert."""
+    run_callback_threadsafe(hass.loop, async_toggle, hass, entity_id)
```
same too

pvizeli commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+        DOMAIN, SERVICE_TOGGLE, async_handle_alert_service,
+        descriptions.get(SERVICE_TOGGLE), schema=ALERT_SERVICE_SCHEMA)
+
+    for alert in all_alerts.values():
```
We are async :) use:

```
tasks = [alert.async_udpate_ha_state() for alert in all_alerts.value()]
if tasks:
    asyncio.wait(tasks, loop=hass.loop)
```

rmkraus replied:
I'm not an expert with asyncio, but, did you intend to be calling the async_update_ha_state method? Did you intend for that to be more like this:

```python
tasks = [alert.async_update_ha_state for alert in all_alerts.values()]
if tasks:
    yield from asyncio.wait(tasks, loop=hass.loop)
```

Otherwise, I think you are just running the update_ha_state methods synchronously. If my presumption is correct (although it may not be), that doesn't work because the Alert class is not hashable. I think, what I have should be relatively asynchronous. However, I am not very well versed with asyncio yet.

pvizeli replied:
You need use `alert.async_update_ha_state()`

rmkraus replied:
I understand now. Thanks.

rmkraus commented:
@kellerza 

If it is any consolation, the messages would actually read more like this:

  - Garage door is open
  - Garage door is open
  - Garage door is open
  - Garage door is open
  - Garage door is open

So I don't think it will trigger anyone's OCD. The only reason, that I can think of, to use lists instead of a factor is for a feeling of unnecessary control. If someone wants to change this implementation later, that's fine. For now, I'm more than satisfied.

The min setting would be for backoff factors less than 1. Some alerts should become more annoying and persistent as they are left unresolved and unacknowledged. A backoff factor less than 1 allows them to do this.

rmkraus commented:
Documentation has been written. Pending final reviews, this is ready for merging.

kellerza commented:
I really think this is implemented without any foresight. You realise this will be a breaking change later?

A list would cater for the `min` case and messages getting more Frequent as well

I will PR the time in the message to trigger your OCD

andrey-git commented:
I think a list is a better approach than a factor.

rmkraus commented:
Respectfully, I disagree. Allow me to explain my thinking on this one:

  1) A factor is much simpler to implement in the code. It is a very simple one-liner that is quite readable. This improved code maintainability.
  2) A factor is much simpler in the configuration file. Many users struggle with proper yaml syntax, especially with deeply embedded values. Granted, there is still a list embedded in the configuration for this component, but adding more would complicate the configuration more.
  3) A factor with a max/min value provides a natural change in delays that can cover most desired behaviors.


A list, on the other hand:

  1) allows very deep control over the delay behavior. However, this level of control is not necessary. It is never necessary for an alert to fire exactly when a user may expect it to. In fact, a user should never be expecting an alert. Alerts notify people of unexpected events so that they may be responded to.
  2) A list allows for a wide range of behaviors, such as delays getting longer and then shorter again. This, I would argue, is outside the 90% of use cases. I see no reason to introduce unnecessarily complicated code to cover the last 10% when a perfectly elegant alternative is available.


All that being said, I don't think an argument can be made that this issue really requires as much discussion as it has received. Both are valid ways of implementing similar logic. This is just the way I did it and those are the reasons I did it that way.

kellerza reviewed (dismissed):
@rmkraus From a user point lists are simpler (4 keys vs 1) and allows more control. Refer to #5215 to see how code complexity can increase for the sake of control,- it's part of progress!

But you are correct that we are spending too much time discussing this. My suggestion is to remove this feature and someone else can a a list implementation without introducing a breaking change for the users.

My apologies for suggesting back-off and complicating this PR

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+            _LOGGER.info('Alerting: %s', self._name)
+            for target in self._notifiers:
+                yield from self.hass.services.async_call(
+                    'notify', target, {'message': self._name})
```
It's confusing that `name` is used for `message`.

rmkraus replied:
Right now the entity's friendly name is what is sent in the message. So it is both, the name and the message. Later, to implement template messages, this could be readdressed.

balloob commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,283 @@
+CONF_CAN_ACK = 'can_acknowledge'
+CONF_NOTIFIERS = 'notifiers'
+CONF_REPEAT = 'repeat'
+CONF_REPEAT_MAX = 'max_repeat'
```
This needs better naming, what about `max_repeat_delay`.

balloob replied:
(in my last comment I suggest to remove this, so let's do that 👍 )

balloob commented:
So I finally had some time to read through all the discussions and code.

 1. The idea of a nagging alerter component is great. It will help people set up this type of automation easily. Although we try to aim for decoupling a lot of logic, for automation-template components a one stop solution is fine. Similar to the `device_sun_light_trigger`.
 2. I like the use case from @kellerza in the beginning more - having it be triggered from automations. This will be possible as we have a `turn_on` service. However, in that case we should be able to operate without an `entity_id` being passed in but that is currently impossible. This could be a future improvement.
 3. The default notifiers are ok but should be extended in a future PR to actually leverage template support for the message. 
 4. Notifiers are ok but I much rather prefer implementing embedded scripts. It will allow any action to be taken, including notifications, and also supports templates. It's what we have been using everywhere. (again, future PR)
 5. Now to repeat, I much prefer a list too. However for a future PR. I think that `backoff`, `max_repeat` and `min_repeat` are giving the user too much control and would only support edge cases for people. Let's just remove it. The people that need such things, a list will be fine too and allows people to exercise even greater control.
_(For the list part, just validate `repeat` with `cv.ensure_list`. Cycle over list for changing repeat interval and stick to last option when options exhausted)_

rmkraus committed:
Changed alert back off to list rather than factor.

rmkraus commented:
Can't argue with the dictator for life 😜. Backoffs are now handled by making the repeat a list. The other features you mentioned, Paulus, could make good additions in future PRs.

rmkraus committed:
Merge branch 'dev' into alert-component

rmkraus commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,271 @@
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_ENTITY_ID): cv.string,
+    vol.Required(CONF_STATE, default=STATE_ON): cv.string,
+    vol.Required(CONF_REPEAT): cv.ensure_list,
```
How can I ensure that this is a list of floats? At the moment, a non-number could be input which would cause issues later on.

rmkraus committed:
Improved async during creation of alerts.

rmkraus commented:
Should be ready for review again.

rmkraus committed:
Improved data type checking for alert component.

kellerza commented:
@rmkraus, thanks! 

Glad to see it as a list and in the end you managed to squeeze the feature in at a 1/5th of the original code length as well. So happy days all around.

rmkraus commented:
Admittedly, it was simpler to implement than I had originally thought. It requires less tests too, which is nice. 

Is everyone ok with merging?

kellerza commented:
I've never reviewed, only commented about the features 😃 

Can review tonight if not done yet

rmkraus commented:
Sure, go ahead and review when you have a chance. Should be done.

kellerza commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,272 @@
+ALERT_SCHEMA = vol.Schema({
+    # pylint: disable=no-value-for-parameter
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_ENTITY_ID): cv.string,
```
Instead of `cv.string` lets do `cv.entity_id`

kellerza commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,272 @@
+        self._firing = False
+        self._ack = False
+        self._cancel = None
+        self.entity_id = ENTITY_ID_FORMAT.format(slugify(entity_id))
```
slugify not required if the schema is `cv.slug`

kellerza commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,272 @@
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        cv.string: ALERT_SCHEMA,
```
Instead of `cv.string` you can require it to be a slug with `cv.slug`. This also makes it compatible with packages. Look at `input_*`, `group`, `automation` etc...

kellerza commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,272 @@
+class Alert(ToggleEntity):
+    """Representation of an alert."""
+
+    def __init__(self, hass, entity_id, name, entity, state, repeat,
```
Let's rename `entity` to `watched_entity_id`

kellerza commented on homeassistant/components/alert.py:
```diff
@@ -0,0 +1,272 @@
+        DOMAIN, SERVICE_TOGGLE, async_handle_alert_service,
+        descriptions.get(SERVICE_TOGGLE), schema=ALERT_SERVICE_SCHEMA)
+
+    tasks = [alert.async_update_ha_state() for alert in all_alerts.values()]
```
Do we need to update the state for an inactive entity?

rmkraus replied:
I think it is best practice. This ensures that the entity is registered with HASS. I don't think there is any downside to doing this and I think it is expected that entities register themselves as soon as they can.

kellerza replied:
Ah, sorry, only now see you push the state everywhere else as well. We normally either follow push/poll model

Can you overwrite [should_poll](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py#L74) to be `False` for your Alert class?

rmkraus replied:
Done. Since these entities do not go into `entity_component`s, the `should_poll` property will never be used, I think. Still, setting `should_poll` to `False` is definitely best practice. Good catch.

kellerza reviewed (commented):
Just some Schema changes, but works well!

homeassistant commented:
Hi @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

rmkraus committed:
Implemented suggested fixes.

homeassistant commented:
Hi @rmkraus, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5201).

Once you do that we will be able to review and accept this pull request.

Thanks!

rmkraus commented:
@kellerza I implemented all the changes your pointed out. Thanks for the feedback. 

I do think that entities should register themselves with HASS as soon as they are created. I'm not married to that though.

Sorry for the delay.

rmkraus committed:
Set the should_poll for alerts to False.

kellerza commented:
Thanks! 🐬
